### ISSUE TITLE
cql-pytest testIndexOnRegularColumnInsertExpiringColumn fails sporadically

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#9008]
In one Jenkins test run, https://jenkins.scylladb.com/job/scylla-master/job/next/3742/consoleText, we saw the following cql-pytest test fail: `cassandra_tests/validation/entities/secondary_index_test.py`::`testIndexOnRegularColumnInsertExpiringColumn`.
This test was translated from Cassandra's unit tests, and tests that TTL works on secondary index. The relevant part of the code is:

>  execute(cql, table, "CREATE INDEX on %s(a)")
>   ...
>  execute(cql, table, "UPDATE %s USING TTL 1 SET a = 5 WHERE pk = 1 AND ck = 2")
>  time.sleep(1.0)
> **assert_empty(execute(cql, table, "SELECT * FROM %s WHERE a = 5"))**

The marked assertion failed because the SELECT returned an unexpired row, not empty.

I ran this test 250 times on my laptop, and it passed every time, so it can't be something like whole-second TTL rounding that would have showed up very often, and has to be more subtle.

As usual, the test machine can be very busy and prone to hangs - but I don't see how this explains anything given that the expiration time of the write is calculated first, and only then the sleep starts. Could Linux's / Python's sleep(1.0) actually sleep less than a full second?.

**Comment by nyh** [comment]
I *was* able to reproduce this bug by running the same test hundreds of more times. The following command runs it 1000 times and stops on the same failure (note that this takes a long time - as each test has two sleeps of one second - parallelism can be used to make it fail faster).
```
test/cql-pytest/run -vs --count=1000 -x cassandra_tests/validation/entities/secondary_index_test.py::testIndexOnRegularColumnInsertExpiringColumn
```
In my test I saw a different failure in the same test, but it's the same underlying problem:
```
            execute(cql, table, "UPDATE %s USING TTL 1 SET b = 10 WHERE pk = 1 A
ND ck = 2")
            time.sleep(1.0)
>           assert_rows(execute(cql, table, "SELECT * FROM %s WHERE a = 3"), [1,
 2, 3, None])
...
E           assert [1, 2, 3, 10] == [1, 2, 3, None]
E             At index 3 diff: 10 != None
E             Full diff:
E             - [1, 2, 3, None]
E             ?           ^^^^
E             + [1, 2, 3, 10]
E             ?           ^^
```

I tested that Python's sleep(1.0) never sleeps for **less** than one second. However, it can come arbitrary close to being over 1.0 seconds - the lowest I saw was 1.0001 (i.e., 0.1ms above 1 second).

My current suspicion is that although expiration-time calculation does not round the time to whole seconds (that would have made this bug much more conspicuous), maybe the way we use gc_clock (seastar::lowres_clock) somehow causes rounding to 1ms (this clock's period) or to 10ms (this clock's actual resolution). I still don't know exactly how or why.

I'm tempted to say that although it would be interesting to understand the exact mechanism why in rare cases a 1 second sleep is not enough for the item to expire, nobody really cares if 0.1ms after one second the item expires or it takes 10ms more. In that case, the test should be modified to sleep 1.01 second (or even 1.1 second, what the heck...).

**Comment by nyh** [comment]
Ok, I know what causes this problem. It is indeed how Seastar's lowres clock works:

Seastar's lowres clock has a "granularity" of 10ms. But it it's not as simple as taking the accurate nanosecond time and truncating it to 10ms resolution. Rather, what Seastar does is to keep a millisecond-resolution "now" variable for the lowres clock, and it sets a timer to wake up roughly every 10ms (`_granularity`), and recalculate this "now". Although this timer **wants** to wake up every 10ms, if this wakeup is delayed for a while (by something else running), for the duration of this delay the time is not yet updated and continues to show the previous now.

So the following scenario is possible:
1. The high-resolution time now is 1.0001, the lowres_clock timer has just run - so we have lowres_clock::now() = 1.000
2. The user writes data with TTL=1. Its expiration time is recorded as 2.
3. The user sleeps for 1.0 seconds. The sleep is from high-resolution time 1.001 to 2.001.
4. After the sleep, the high-resolution time is indeed 2.001, but because it's not guaranteed when exactly the lowres timer ran, it might have run, say, 5ms ago, when the time was 1.996.
5. So Scylla thinks the lowres time is now still 1.996, which is <2, so the item is not considered expired!

I confirmed that if I increase the sleep to 1.1, the test passes even 2000 runs. So I'll create a patch to do that.

**Comment by nyh** [comment]
No backports needed - the flaky test fixed here was a new test, added in commit ff81072f64c65156c05fa0bb1eddbde18e5602dd and never reached any release.
